src/Workspaces/Core/Portable/Options/SolutionOptionSet.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Xml.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    /// <summary>
+    /// An implementation of <see cref="OptionSet"/> that fetches values it doesn't know about to the workspace's option service. It ensures a contract
+    /// that values are immutable from this instance once observed.
+    /// </summary>
+    internal sealed class SolutionOptionSet : SerializableOptionSet
+    {
+        private readonly WorkspaceOptionSet _workspaceOptionSet;
```

Comment:
NOTE: Unfortunately, `SolutionOptionSet` is still not purely data/snapshot and needs the fallback `WorkspaceOptionSet`  due to the fact that large number of our options and option providers are defined in layers above Features (i.e. EditorFeatures and VS layer), which are not available in remote workspace. We need to ensure that accessing OR updating these options through `Solution.Options` or `Workspace.Options` is equivalent in VS, so this field is required. Once we move all the option providers down to Workspaces or Features, we can remove this field and this option set can become pure data.

Note that this PR still takes us to a better place then before - earlier no options were synced to remote workspace, unless the feature invoking the remote APIs itself explicitly passed option set as an additional asset. Now all the options defined in Workspace and Features layer are always part of solution snapshot and available in remote workspace.